API.md

```diff
@@ -0,0 +1,93 @@
+# Introduction
+This document provides API specification of NodeResourceTopology API which is a CRD
+based API that describes the structure of topology on a node. One of the goals when
+designing this API was to ensure hardware independent data structure, which allows
+us to have a future-proof API that represents hardware of varying internal topology.
+
+CRD instances must be created external to kubelet and independently. Each instance
+will be bound to one working node.
+
+# NodeTopology Object Definition
+
+```
+// NodeResourceTopology describes node resources and their topology.
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Zones      ZoneList      `json:"zones"`
+	Attributes AttributeList `json:"attributes,omitempty"`
+}
+
+// Zone represents a resource topology zone, e.g. socket, node, die or core.
+type Zone struct {
+	Name       string           `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type       string           `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Parent     string           `json:"parent,omitempty" protobuf:"bytes,3,opt,name=parent"`
+	Costs      CostList         `json:"costs,omitempty" protobuf:"bytes,4,rep,name=costs"`
```

Comment:
Is Zone the best term to describe a resource group? Zone already has a specific term in cloud (e.g., east  / west  usually a location).

Reply:
> Will the idea here be that costs are compared with some Less (or similar function) to determine an optimal placement?

Yes, the cost basically represents the "topological distance" between zones. Or the cost of moving data between them.

> I'm guessing cost is akin to an edge - is that correct?

Yes. It can be thought as the weight of an edge in a weighed graph.

> If this is a graph is there any reason to not model it more explicitly as one (naming-wise)?

I think the naming was chosen for the context (hardware topology) instead of using just generic abstract naming of graphs. Any counterproposals?

> And there are multiple in the list to allow for different kinds of costs (e.g., different edges - a node to node cost vs. node to something else?)

It's not for different types of costs but cost to different zones. If you have 4 zones A, B, C and D you want cost between each of them, e.g. for A: A->B, A->C and A->D

> Why was a list chosen instead of a map for easy lookup?

I think one reason is the k8s API conventions/guidelines to prefer lists of named objects over maps.

> Is Zone the best term to describe a resource group? I know it has an explicit meaning for hardware, but in the context of cloud native folks we will think of something else. Is there a qualifier we can add?

It's not supposed to be just some "random" resource group. The assumption kinda is that it's some topologically relevant group. So I think zone is good but I'm open to hear other opinions.

Anything to correct or add @swatisehgal @ffromani ?

Reply:
Agree with everything and still on the lookout to find a reference that k8s APIs prefer lists over maps, because I also clearly remember that suggestions but can't find it atm.

Reply:
Agree with everything here too. Thanks @marquiz!

> > Why was a list chosen instead of a map for easy lookup?
> 
> I think one reason is the k8s API conventions/guidelines to prefer lists of named objects over maps.

Yep, that's correct. [Here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) is the reference to API conventions because of which we updated this API to use lists instead of maps.
 
> > Is Zone the best term to describe a resource group? I know it has an explicit meaning for hardware, but in the context of cloud native folks we will think of something else. Is there a qualifier we can add?
> 
> It's not supposed to be just some "random" resource group. The assumption kinda is that it's some topologically relevant group. So I think zone is good but I'm open to hear other opinions.

Agreed, we were aiming for an umbrella term that can be used to represent a range of topological zones in hardware and the term Zone seemed to fit the bill. From cloud native context perspective, if you have better naming suggestions we would be happy to consider.

Reply:
> range of topological zones in hardware 

Do you have a document or reference that talks about hardware zones? I'm new to the idea (and want to brainstorm a possible other term, but perhaps there is not a good replacement)!

Reply:
The idea is that a zone could be any topological zone like core, NUMA, package, die, node etc. We have example in [this](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit?usp=sharing) that might help.